File: src/baekjoon/graph/b11657/b11657.py
import sys

input = sys.stdin.readline

INF = float('inf')

def solve():
    N, M = map(int, input().split()) # N: 도시수, M: 버스 노석 수
    edges = []
    for _ in range(M):
        A, B, C = map(int, input().split())
        edges.append((A, B, C)) # (출발, 도착, 시간)

    # 1. 거리 리스트 초기화
    dist = [INF] * (N + 1)
    dist[1] = 0 # 시작 도시(1)의 거리는 0

    # 2. N-1 번 동안 모든 간선에 대한 완화 작업 반복
    # 추가 설명: 한번 돌때는 경로 1(간선 1개를 이용해)로 완성된 최단 경로 찾음
    # 두번 돌때는 경로 2(간선 2개를 이용해)로 완성된 최단 경로 찾음...
    # N-1번 돌때는 경로 N-1(간선 N-1개를 이용해)로 완성된 최단 경로 찾음
    for i in range(N-1):
        for u, v, w in edges:
            # 출발점이 도달 가능한 상태이고, 현재 간성을 거치는 것이 더 빠른 경우
            if dist[u] != INF and dist[v] > dist[u] + w:
                dist[v] = dist[u] + w

    # 3. 음수 사이클 확인 (N번쨰 완화 작업)
    has_negative_cycle = False
    for u, v, w in edges:
        # N -1 번 반복 후에도 거리가 갱신된다면 음수 사이클 존재
        if dist[u] != INF and dist[v] > dist[u] + w:
            has_negative_cycle = True
            break

    # 음수 사이클이 존재하면 -1 출력
    if has_negative_cycle:
        print(-1)
    else:
        # 4. 결과 출력
        for i in range(2, N+1):
            if dist[i] == INF:
                print(-1)
            else:
                print(dist[i])
solve()

File: src/baekjoon/graph/b11657/b11657.java
package baekjoon.graph.b11657;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

// 간선 정보를 담을 클래스
class Edge {
    int from;
    int to;
    int cost;

    public Edge(int from, int to, int cost) {
        this.from = from;
        this.to = to;
        this.cost = cost;
    }
}

public class b11657 {
    static final long INF = Long.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        List<Edge> edges = new ArrayList<>();
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());
            edges.add(new Edge(from, to, cost));
        }

        // 1. 거리 배열 초기화
        long[] dist = new long[N + 1];
        Arrays.fill(dist, INF);
        dist[1] = 0; // 시작 도시(1)의 거리는 0

        // 2. N-1번 동안 모든 간선에 대한 완화 작업 반복
        for (int i = 0; i < N - 1; i++) {
            for (Edge edge : edges) {
                if (dist[edge.from] != INF && dist[edge.to] > dist[edge.from] + edge.cost) {
                    dist[edge.to] = dist[edge.from] + edge.cost;
                }
            }
        }

        // 3. 음수 사이클 확인(N 번쨰 완화 작업)
        boolean hasNegativeCycle = false;
        for (Edge edge : edges) {
            // N - 1번 반복 후에도 거리가 갱신된다면 음수 사이클 존재
            if (dist[edge.from] != INF && dist[edge.to] > dist[edge.from] + edge.cost) {
                hasNegativeCycle = true;
                break;
            }
        }
        StringBuilder sb = new StringBuilder();
        if (hasNegativeCycle) {
            sb.append(-1).append("\n");
        } else {
            for (int i = 2; i <= N; i++) {
                if (dist[i] == INF) {
                    sb.append(-1).append("\n"); // 도달 불가
                } else {
                    sb.append(dist[i]).append("\n");
                }
            }
        }
        System.out.print(sb);
    }
}
